x-pack/plugin/stack/src/javaRestTest/java/org/elasticsearch/xpack/stack/EcsDynamicTemplatesIT.java

```diff
@@ -0,0 +1,416 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.stack;
+
+import org.apache.http.HttpStatus;
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.network.NetworkAddress;
+import org.elasticsearch.common.time.DateFormatter;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.test.cluster.ElasticsearchCluster;
+import org.elasticsearch.test.rest.ESRestTestCase;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xcontent.json.JsonXContent;
+import org.elasticsearch.xpack.core.template.TemplateUtils;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+
+@SuppressWarnings("unchecked")
+public class EcsDynamicTemplatesIT extends ESRestTestCase {
+
+    @ClassRule
+    public static ElasticsearchCluster cluster = ElasticsearchCluster.local().module("mapper-extras").module("wildcard").build();
+
+    // The dynamic templates we test against
+    public static final String ECS_DYNAMIC_TEMPLATES_FILE = "ecs-dynamic-mappings.json";
+
+    // The current ECS state (branch main) containing all fields in flattened form
+    private static final String ECS_FLAT_FILE_URL = "https://raw.githubusercontent.com/elastic/ecs/main/generated/ecs/ecs_flat.yml";
```

Comment:
@rjernst noted that the problem with fetching from main means that once an ECS change breaks this test, all PRs will fail until we get the dynamic templates (or the test) fixed. @mark-vieira raised a similar [concern](https://github.com/elastic/elasticsearch/issues/96713#issuecomment-1603760192).
I see why relying on main is problematic, but I am not sure what the alternative is. 

If we fetch the `ecs_flat.yml` file from a version branch, for example, it means that we can only capture ECS mismatches after it was already released, which implies that we will ship at least one version with dynamic templates mismatching the ECS of the same version (could be more if we rely on manually updating the used version branch and we fail to do that frequently enough).

I think that a state where `ecs-dynamic-mappings.json` doesn't fully cover ECS should be viewed as a bug. Relying on main provides a good way to ensure that we always release both in sync.

#96713 originally proposed that these tests will run **periodically** but I got advices to add them as yet another test in the suite. Maybe we need to reconsider periodic testing of this?

Any proposals or feedback on that are very welcome.

Reply:
We should have any changes to ECS run these tests so we can avoid this scenario as much as possible. Testing this unilaterally means that any breakages or disruptions are on the Elasticsearch side and that's not ideal. In theory, if every change on either end runs these tests, this scenario should never happen.

Reply:
Can you assist with setting this up? 
Alternatively, it may be much simpler to run this test daily/weekly. Should be good enough in order to catch inconsistencies soon enough after they are made to ECS main without affecting any other CI run in Elasticsearch.

Reply:
@mark-vieira how should we proceed with this? Would scheduled daily/weekly builds work?

Reply:
Where does CI for ECS live currently?

Reply:
I am not familiar with ECS CI, maybe @ebeahan can assist with that.
This is open for quite some time, what would you say is the best way forward?